src/tech/v3/protocols/column.clj

```diff
@@ -45,3 +45,23 @@ Supported types are:
     "Convert to a java primitive array of a given datatype.  For strings,
 an implicit string->double mapping is expected.  For booleans, true=1 false=0.
 Finally, any missing values should be indicated by a NaN of the expected type."))
+
+
+(defprotocol PHasIndexStructure
+  (index-structure [this]
+    "Returns an index-structure for the column. The index-structure will only be
+generated the first time this is function is called.")
+  (index-structure-realized? [this]
+    "Returns true if the index-structure value has been produced. The index-structure
+is only produced the first time it is requested.")
```

Comment:
@cnuernber Is this what you had in mind [here](https://github.com/techascent/tech.ml.dataset/pull/214#pullrequestreview-645095373) when you said that a fn like `filter-columns` might want to check if a column had an "index" assigned? I initially interpreted that as an optimization thing. I.e. when calling filter-column on a column, we might not want to use the index if it hadn't already been generated since there might be a faster way? But then reading it over again I wasn't so sure. Since if there's a faster way to work on the column than the index-structure why would we have the index at all? So perhaps I misunderstood...

Reply:
Implementation here: https://github.com/techascent/tech.ml.dataset/pull/214/files#diff-63af69d2c32736df75a6e42b11870179aee860900769c0b261c338666ebd23ceR168-R169

Reply:
You have it right.  When calling filter-column, depending on the 'predicate' passed in we want to use the index structure if it is available but if not we just want to use the (parallelized) original version of [argfilter](https://github.com/cnuernber/dtype-next/blob/master/src/tech/v3/datatype/argops.clj#L442).

Reply:
That is true in general of most queries to the dataset I think.  If properly constructed the index structure should simply be an optimization to the query.